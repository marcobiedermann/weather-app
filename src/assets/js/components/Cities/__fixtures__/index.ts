import cityFixtures from '../../City/__fixtures__';

const cities = [cityFixtures];

export default cities;
