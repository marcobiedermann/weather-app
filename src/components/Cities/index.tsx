import React from 'react';
import CityPreview, { CityPreviewProps } from '../CityPreview';
import styles from './style.module.css';

interface CitiesProps {
  cities?: CityPreviewProps[];
}

function Cities(props: CitiesProps): JSX.Element {
  const { cities = [], ...otherProps } = props;

  return (
    <ul className={styles.cities} {...otherProps}>
      {cities.map((city) => (
        <li key={city.id}>
          <CityPreview {...city} />
        </li>
      ))}
    </ul>
  );
}

export type { CitiesProps };
export default Cities;
