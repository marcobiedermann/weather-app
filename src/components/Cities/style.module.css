.cities {
  display: grid;
  gap: 1.5em;
  list-style: none;
  margin-block: 0.5em 1.5em;
  padding: 0;
}
