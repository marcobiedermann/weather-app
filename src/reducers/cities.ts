import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';

interface CitiesState {
  cities: number[];
}

const initialState: CitiesState = {
  cities: [],
};

const citiesSlice = createSlice({
  name: 'cities',
  initialState,
  reducers: {
    addCity: (state, action: PayloadAction<number>) => {
      state.cities.push(action.payload);
    },
    removeCity: (state, action: PayloadAction<number>) => {
      state.cities = state.cities.filter((city) => city !== action.payload);
    },
  },
});

const { addCity, removeCity } = citiesSlice.actions;

const selectCities = (state: RootState): number[] => state.cities.cities;

export { addCity, removeCity, selectCities };
export default citiesSlice.reducer;
