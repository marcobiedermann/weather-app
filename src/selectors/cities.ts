import type { RootState } from '../store';

function selectCities(state: RootState): number[] {
  return state.cities.cities;
}

export { selectCities };
