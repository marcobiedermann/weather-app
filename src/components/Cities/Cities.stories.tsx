import type { Meta, StoryObj } from '@storybook/react';
import Cities from '.';

const meta: Meta<typeof Cities> = {
  component: Cities,
  args: {
    cities: [
      {
        sys: {
          country: 'DE',
        },
        main: {
          temp: 18.54,
        },
        id: 2950159,
        name: 'Berlin',
      },
    ],
  },
};

type Story = StoryObj<typeof Cities>;

const Default: Story = {};

export { Default };
export default meta;
