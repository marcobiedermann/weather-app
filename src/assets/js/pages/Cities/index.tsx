import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { Link, useRouteMatch } from 'react-router-dom';
import useSWR from 'swr';
import Cities from '../../components/Cities';
import Error from '../../components/Error';
import Loader from '../../components/Loader';
import { selectCities } from '../../reducers/cities';

interface CitiesPageQuery {
  cnt: number;
  list: List[];
}

interface List {
  coord: Coord;
  sys: Sys;
  weather: Weather[];
  main: Main;
  visibility: number;
  wind: Wind;
  clouds: Clouds;
  dt: number;
  id: number;
  name: string;
}

interface Clouds {
  all: number;
}

interface Coord {
  lon: number;
  lat: number;
}

interface Main {
  temp: number;
  feels_like: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
}

interface Sys {
  country: string;
  timezone: number;
  sunrise: number;
  sunset: number;
}

interface Weather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

interface Wind {
  speed: number;
  deg: number;
}

const CitiesPage: FC = () => {
  const cities = useSelector(selectCities);
  const { path } = useRouteMatch();
  const { t } = useTranslation();
  const { data, error } = useSWR<CitiesPageQuery>(`/group?id=${cities.join(',')}`);

  if (error) {
    return <Error message={error.message} />;
  }

  if (!data) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>Cities</title>
      </Helmet>
      <Cities cities={data.list} />
      <Link to={`${path}/edit`}>{t('translation:edit')}</Link>
    </>
  );
};

export default CitiesPage;
