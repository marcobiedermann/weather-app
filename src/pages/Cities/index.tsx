import { useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';
import Cities from '../../components/Cities';
import Error from '../../components/Error';
import Link from '../../components/Link';
import Loader from '../../components/Loader';
import Seo from '../../components/Seo';
import { useGroup } from '../../hooks';
import { selectCities } from '../../selectors/cities';
import { selectSettings } from '../../selectors/settings';
import { useAppSelector } from '../../store';

function CitiesPage(): JSX.Element {
  const { pathname } = useLocation();
  const navigate = useNavigate();
  const cities = useAppSelector(selectCities);
  const settings = useAppSelector(selectSettings);
  const { t } = useTranslation();
  const { data: group, error, isError, isPending } = useGroup(cities, settings);

  if (!cities.length) {
    navigate('/search');
  }

  if (isPending) {
    return <Loader />;
  }

  if (isError) {
    return <Error message={error.message} />;
  }

  return (
    <>
      <Seo title="Cities" />
      <Cities cities={group.list} />
      <Link to={`${pathname}/edit`}>{t('edit')}</Link>
    </>
  );
}

export default CitiesPage;
