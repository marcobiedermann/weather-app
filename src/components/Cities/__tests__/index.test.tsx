import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { describe, expect, it } from 'vitest';
import Cities from '..';
import cities from '../__fixtures__';

describe('components/Cities', () => {
  it('should render correctly', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Cities cities={cities} />
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render empty cities correctly', () => {
    const { asFragment } = render(<Cities cities={[]} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
