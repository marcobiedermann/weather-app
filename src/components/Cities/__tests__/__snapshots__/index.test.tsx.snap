// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/Cities > should render correctly 1`] = `
<DocumentFragment>
  <ul
    class="cities"
  >
    <li>
      <a
        class="city"
        href="/cities/2950159"
      >
        <div>
          <h3
            class="city__country"
          >
            DE
          </h3>
          <h2
            class="city__name"
          >
            Berlin
          </h2>
        </div>
        <div
          class="city__temp"
        >
          12.61 Â°C
        </div>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`components/Cities > should render empty cities correctly 1`] = `<DocumentFragment />`;
